velox/exec/HashProbe.cpp

```diff
@@ -189,58 +222,278 @@ void HashProbe::initializeFilter(
+  }
+
+  // If 'spillInputPartitionIds_' is not empty, then we set up a spiller to
+  // spill the incoming probe inputs.
+  const auto& spillConfig = spillConfig_.value();
+  spiller_ = std::make_unique<Spiller>(
+      Spiller::Type::kHashJoinProbe,
+      probeType_,
+      HashBitRange(
+          spillInputPartitionIds_.begin()->partitionBitOffset(),
+          spillInputPartitionIds_.begin()->partitionBitOffset() +
+              spillConfig.hashBitRange.numBits()),
+      spillConfig.filePath,
       operatorCtx_->task()
-          ->getHashJoinBridge(
-              operatorCtx_->driverCtx()->splitGroupId, planNodeId())
-          ->tableOrFuture(future);
+              ->queryCtx()
+              ->pool()
+              ->getMemoryUsageTracker()
+              ->maxTotalBytes() *
+          spillConfig.fileSizeFactor,
+      Spiller::spillPool(),
+      spillConfig.executor);
+  // Set the spill partitions to the corresponding ones at the build side. The
+  // hash probe operator itself won't trigger any spilling.
+  spiller_->setPartitionsSpilled(toPartitionNumSet(spillInputPartitionIds_));
+
+  spillHashFunction_ = std::make_unique<HashPartitionFunction>(
+      spiller_->hashBits(), probeType_, keyChannels_);
+  spillInputIndicesBuffers_.resize(spillHashFunction_->numPartitions());
+  rawSpillInputIndicesBuffers_.resize(spillHashFunction_->numPartitions());
+  numSpillInputs_.resize(spillHashFunction_->numPartitions(), 0);
+}
+
+void HashProbe::waitForHashTable() {
+  checkRunning();
+  VELOX_CHECK_NULL(table_);
+
+  auto hashBuildResult = joinBridge_->tableOrFuture(&future_);
   if (!hashBuildResult.has_value()) {
-    VELOX_CHECK_NOT_NULL(future);
-    return BlockingReason::kWaitForJoinBuild;
+    VELOX_CHECK(future_.valid());
+    setState(State::kWaitForBuild);
+    return;
   }
 
   if (hashBuildResult->antiJoinHasNullKeys) {
-    // Anti join with null keys on the build side always returns nothing.
+    // Null-aware anti join with null keys on the build side without a filter
+    // always returns nothing.
     VELOX_CHECK(isNullAwareAntiJoin(joinType_));
-    finished_ = true;
-  } else {
-    table_ = hashBuildResult->table;
-    if (table_->numDistinct() == 0) {
-      // Build side is empty. Inner, right and semi joins return nothing in this
-      // case, hence, we can terminate the pipeline early.
-      if (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
-          isRightJoin(joinType_) || isRightSemiJoin(joinType_)) {
-        finished_ = true;
+    // 'antiJoinHasNullKeys' flag shall only be set on the first built 'table_'
+    VELOX_CHECK(spillPartitionSet_.empty());
+    noMoreInput();
+    return;
+  }
+
+  table_ = std::move(hashBuildResult->table);
+  VELOX_CHECK_NOT_NULL(table_);
+
+  maybeSetupSpillInput(
+      hashBuildResult->restoredPartitionId, hashBuildResult->spillPartitionIds);
+
+  if (table_->numDistinct() == 0) {
+    // If the build table is empty, then inner, right and semi joins return
```

Comment:
nit: skipProbeOnEmptyBuild is a nice descriptive name that doesn't seem to require a comment; would it make sense to move this comment to the header file to shorten this function?